File: app/src/main/java/com/example/puffer/parkingdemo/model/repository/HistoryDao.java
package com.example.puffer.parkingdemo.model.repository;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import com.example.puffer.parkingdemo.model.data.History;
import com.example.puffer.parkingdemo.model.data.Park;

import io.reactivex.Completable;
import io.reactivex.Single;

@Dao
public interface HistoryDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    Completable insert(History item);

    @Query("SELECT Table_Parking.* FROM Table_History INNER JOIN Table_Parking " +
            "ON Table_History.id = Table_Parking.id ORDER BY Table_History.hashTag DESC")
    Single<Park[]> getHistoryList();

    @Query("DELETE FROM Table_History WHERE id LIKE :id")
    Completable deleteByID(String id);
}

File: app/src/main/java/com/a1573595/parkingdemo/data/local/HistoryDao.kt
package com.a1573595.parkingdemo.data.local

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Query
import androidx.room.Transaction
import androidx.room.Upsert
import com.a1573595.parkingdemo.domain.model.ParkingLot
import kotlinx.coroutines.flow.Flow

@Dao
interface HistoryDao {
    @Query("SELECT * FROM History INNER JOIN ParkingLot ON History.id = ParkingLot.id Order By History.rowid DESC")
    fun getParkingLotListFlow(): Flow<List<ParkingLot>>

    @Upsert
    suspend fun upsert(history: History)

    @Delete
    suspend fun delete(history: History)

    @Transaction
    suspend fun deleteAndUpsert(history: History) {
        delete(history)
        upsert(history)
    }
}